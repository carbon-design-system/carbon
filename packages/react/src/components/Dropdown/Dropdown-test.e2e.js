/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import '../../../index.scss';

import React from 'react';
import { mount } from '@cypress/react';
import { generateItems, generateGenericItem } from '../ListBox/test-helpers';
import Dropdown from '../Dropdown';
import DropdownSkeleton from '../Dropdown/Dropdown.Skeleton';

describe('Dropdown', () => {
  beforeEach(() => {
    const items = generateItems(5, generateGenericItem);
    const label = 'Dropdown menu options';
    const style = { marginBottom: '1rem' };
    mount(
      <>
        <Dropdown
          style={{ marginBottom: '14rem' }}
          items={items}
          label={label}
        />
        <Dropdown style={style} items={items} label={label} size="sm" />
        <Dropdown style={style} items={items} label={label} size="md" />
        <Dropdown style={style} items={items} label={label} size="lg" />
        <Dropdown style={style} items={items} label={label} size="xl" />
        <Dropdown style={style} items={items} label={label} disabled />
        <Dropdown
          style={style}
          items={items}
          label={label}
          helperText="This is helper text"
        />
        <Dropdown style={style} items={items} label={label} warn />
        <Dropdown
          style={style}
          items={items}
          label={label}
          warn
          warnText="This is warn text"
        />
        <Dropdown
          style={style}
          items={items}
          label={label}
          invalid
          invalidText="This is invalid text"
        />
        <Dropdown style={style} items={items} label={label} light />
        <Dropdown
          style={style}
          items={items}
          label={label}
          light
          invalid
          invalidText="This is invalid text"
        />
        <Dropdown style={style} items={items} label={label} type="inline" />
        <DropdownSkeleton style={style} size="sm" />
        <DropdownSkeleton style={style} size="md" />
        <DropdownSkeleton style={style} size="lg" />
        <DropdownSkeleton style={style} size="xl" />
      </>
    );
  });

  it('should render', () => {
    cy.findAllByText(/Dropdown menu options/)
      .should('have.length', 13)
      .last()
      .should('be.visible');

    // snapshots should always be taken _after_ an assertion that
    // a element/component should be visible. This is to ensure
    // the DOM has settled and the element has fully loaded.
    cy.percySnapshot();
  });

  it('should render listbox when clicked', () => {
    cy.findAllByText(/Dropdown menu options/)
      .first()
      .click();

    cy.findAllByText(/Item 0/)
      .first()
      .should('be.visible');
    cy.findAllByText(/Item 4/)
      .first()
      .should('be.visible');

    // snapshots should always be taken _after_ an assertion that
    // a element/component should be visible. This is to ensure
    // the DOM has settled and the element has fully loaded.
    cy.percySnapshot();
  });
});
