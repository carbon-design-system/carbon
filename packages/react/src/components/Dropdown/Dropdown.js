/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { useRef } from 'react';
import { useSelect } from 'downshift';
import { settings } from 'carbon-components';
import cx from 'classnames';
import PropTypes from 'prop-types';
import {
  Checkmark16,
  WarningAltFilled16,
  WarningFilled16,
} from '@carbon/icons-react';
import ListBox, { PropTypes as ListBoxPropTypes } from '../ListBox';
import { mapDownshiftProps } from '../../tools/createPropAdapter';
import mergeRefs from '../../tools/mergeRefs';
import deprecate from '../../prop-types/deprecate';

const { prefix } = settings;

const defaultItemToString = (item) => {
  if (typeof item === 'string') {
    return item;
  }

  return item ? item.label : '';
};

const Dropdown = React.forwardRef(function Dropdown(
  {
    className: containerClassName,
    disabled,
    direction,
    items,
    label,
    ariaLabel,
    itemToString,
    itemToElement,
    type,
    size,
    onChange,
    id,
    titleText,
    hideLabel,
    helperText,
    translateWithId,
    light,
    invalid,
    invalidText,
    warn,
    warnText,
    initialSelectedItem,
    selectedItem: controlledSelectedItem,
    downshiftProps,
    ...other
  },
  ref
) {
  const selectProps = mapDownshiftProps({
    ...downshiftProps,
    items,
    itemToString,
    initialSelectedItem,
    onSelectedItemChange,
  });

  // only set selectedItem if the prop is defined. Setting if it is undefined
  // will overwrite default selected items from useSelect
  if (controlledSelectedItem !== undefined) {
    selectProps.selectedItem = controlledSelectedItem;
  }

  const {
    isOpen,
    getToggleButtonProps,
    getLabelProps,
    getMenuProps,
    getItemProps,
    highlightedIndex,
    selectedItem,
  } = useSelect(selectProps);
  const inline = type === 'inline';
  const showWarning = !invalid && warn;

  const className = cx(`${prefix}--dropdown`, containerClassName, {
    [`${prefix}--dropdown--invalid`]: invalid,
    [`${prefix}--dropdown--warning`]: showWarning,
    [`${prefix}--dropdown--open`]: isOpen,
    [`${prefix}--dropdown--inline`]: inline,
    [`${prefix}--dropdown--disabled`]: disabled,
    [`${prefix}--dropdown--light`]: light,
    [`${prefix}--dropdown--${size}`]: size,
    [`${prefix}--list-box--up`]: direction === 'top',
  });

  const titleClasses = cx(`${prefix}--label`, {
    [`${prefix}--label--disabled`]: disabled,
    [`${prefix}--visually-hidden`]: hideLabel,
  });

  const helperClasses = cx(`${prefix}--form__helper-text`, {
    [`${prefix}--form__helper-text--disabled`]: disabled,
  });

  const wrapperClasses = cx(
    `${prefix}--dropdown__wrapper`,
    `${prefix}--list-box__wrapper`,
    {
      [`${prefix}--dropdown__wrapper--inline`]: inline,
      [`${prefix}--list-box__wrapper--inline`]: inline,
      [`${prefix}--dropdown__wrapper--inline--invalid`]: inline && invalid,
      [`${prefix}--list-box__wrapper--inline--invalid`]: inline && invalid,
    }
  );

  // needs to be Capitalized for react to render it correctly
  const ItemToElement = itemToElement;
  const toggleButtonProps = getToggleButtonProps();
  const helper = helperText ? (
    <div className={helperClasses}>{helperText}</div>
  ) : null;

  function onSelectedItemChange({ selectedItem }) {
    if (onChange) {
      onChange({ selectedItem });
    }
  }

  const menuItemOptionRefs = useRef(items.map((_) => React.createRef()));

  return (
    <div className={wrapperClasses} {...other}>
      {titleText && (
        <label className={titleClasses} {...getLabelProps()}>
          {titleText}
        </label>
      )}
      <ListBox
        aria-label={ariaLabel}
        size={size}
        className={className}
        invalid={invalid}
        invalidText={invalidText}
        warn={warn}
        warnText={warnText}
        light={light}
        isOpen={isOpen}
        id={id}>
        {invalid && (
          <WarningFilled16 className={`${prefix}--list-box__invalid-icon`} />
        )}
        {showWarning && (
          <WarningAltFilled16
            className={`${prefix}--list-box__invalid-icon ${prefix}--list-box__invalid-icon--warning`}
          />
        )}
        <button
          type="button"
          className={`${prefix}--list-box__field`}
          disabled={disabled}
          aria-disabled={disabled}
          {...toggleButtonProps}
          ref={mergeRefs(toggleButtonProps.ref, ref)}>
          <span className={`${prefix}--list-box__label`}>
            {selectedItem ? itemToString(selectedItem) : label}
          </span>
          <ListBox.MenuIcon isOpen={isOpen} translateWithId={translateWithId} />
        </button>
        <ListBox.Menu {...getMenuProps()}>
          {isOpen &&
            items.map((item, index) => {
              const itemProps = getItemProps({ item, index });
              const title = itemToElement ? item.text : itemToString(item);
              const { offsetWidth, scrollWidth } =
                menuItemOptionRefs?.current[index]?.current || {};
              return (
                <ListBox.MenuItem
                  key={itemProps.id}
                  isActive={selectedItem === item}
                  isHighlighted={
                    highlightedIndex === index || selectedItem === item
                  }
                  title={(offsetWidth < scrollWidth && title) || undefined}
                  ref={{
                    menuItemOptionRef: menuItemOptionRefs.current[index],
                  }}
                  {...itemProps}>
                  {itemToElement ? (
                    <ItemToElement key={itemProps.id} {...item} />
                  ) : (
                    itemToString(item)
                  )}
                  {selectedItem === item && (
                    <Checkmark16
                      className={`${prefix}--list-box__menu-item__selected-icon`}
                    />
                  )}
                </ListBox.MenuItem>
              );
            })}
        </ListBox.Menu>
      </ListBox>
      {!inline && !invalid && !warn && helper}
    </div>
  );
});

Dropdown.displayName = 'Dropdown';
Dropdown.propTypes = {
  /**
   * 'aria-label' of the ListBox component.
   */
  ariaLabel: PropTypes.string,

  /**
   * Provide a custom className to be applied on the bx--dropdown node
   */
  className: PropTypes.string,

  /**
   * Specify the direction of the dropdown. Can be either top or bottom.
   */
  direction: PropTypes.oneOf(['top', 'bottom']),

  /**
   * Disable the control
   */
  disabled: PropTypes.bool,

  /**
   * Additional props passed to Downshift
   */
  downshiftProps: PropTypes.object,

  /**
   * Provide helper text that is used alongside the control label for
   * additional help
   */
  helperText: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),

  /**
   * Specify whether the title text should be hidden or not
   */
  hideLabel: PropTypes.bool,

  /**
   * Specify a custom `id`
   */
  id: PropTypes.string.isRequired,

  /**
   * Allow users to pass in an arbitrary item or a string (in case their items are an array of strings)
   * from their collection that are pre-selected
   */
  initialSelectedItem: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.string,
  ]),

  /**
   * Specify whether you want the inline version of this control
   */
  inline: deprecate(
    PropTypes.bool,
    `The \`inline\` prop has been deprecated and will
    be removed in the next major release. To specify the inline variant of Dropdown, please use the \`type\` prop.`
  ),

  /**
   * Specify if the currently selected value is invalid.
   */
  invalid: PropTypes.bool,

  /**
   * Message which is displayed if the value is invalid.
   */
  invalidText: PropTypes.node,

  /**
   * Function to render items as custom components instead of strings.
   * Defaults to null and is overriden by a getter
   */
  itemToElement: PropTypes.func,

  /**
   * Helper function passed to downshift that allows the library to render a
   * given item to a string label. By default, it extracts the `label` field
   * from a given item to serve as the item label in the list.
   */
  itemToString: PropTypes.func,

  /**
   * We try to stay as generic as possible here to allow individuals to pass
   * in a collection of whatever kind of data structure they prefer
   */
  items: PropTypes.array.isRequired,

  /**
   * Generic `label` that will be used as the textual representation of what
   * this field is for
   */
  label: PropTypes.node.isRequired,

  /**
   * `true` to use the light version.
   */
  light: PropTypes.bool,

  /**
   * `onChange` is a utility for this controlled component to communicate to a
   * consuming component what kind of internal state changes are occuring.
   */
  onChange: PropTypes.func,

  /**
   * In the case you want to control the dropdown selection entirely.
   */
  selectedItem: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),

  /**
   * Specify the size of the ListBox. Currently supports either `sm`, `md` or `lg` as an option.
   */
  size: ListBoxPropTypes.ListBoxSize,

  /**
   * Provide the title text that will be read by a screen reader when
   * visiting this control
   */
  titleText: PropTypes.node,

  /**
   * Callback function for translating ListBoxMenuIcon SVG title
   */
  translateWithId: PropTypes.func,

  /**
   * The dropdown type, `default` or `inline`
   */
  type: ListBoxPropTypes.ListBoxType,

  /**
   * Specify whether the control is currently in warning state
   */
  warn: PropTypes.bool,

  /**
   * Provide the text that is displayed when the control is in warning state
   */
  warnText: PropTypes.node,
};

Dropdown.defaultProps = {
  disabled: false,
  type: 'default',
  itemToString: defaultItemToString,
  itemToElement: null,
  light: false,
  titleText: '',
  helperText: '',
  direction: 'bottom',
};

export default Dropdown;
