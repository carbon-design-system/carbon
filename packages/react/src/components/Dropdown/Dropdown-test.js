/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { cleanup, render } from '@testing-library/react';
import React from 'react';
import { mount, shallow } from 'enzyme';
import {
  assertMenuOpen,
  assertMenuClosed,
  findMenuItemNode,
  openMenu,
  generateItems,
  generateGenericItem,
} from '../ListBox/test-helpers';
import Dropdown from '../Dropdown';
import DropdownSkeleton from '../Dropdown/Dropdown.Skeleton';
import { settings } from 'carbon-components';

const { prefix } = settings;

describe('Dropdown', () => {
  let mockProps;
  beforeEach(() => {
    mockProps = {
      id: 'test-dropdown',
      items: generateItems(5, generateGenericItem),
      onChange: jest.fn(),
      label: 'input',
      placeholder: 'Filter...',
      type: 'default',
    };
  });

  it('should render', () => {
    const wrapper = mount(<Dropdown {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should initially render with the menu not open', () => {
    const wrapper = mount(<Dropdown {...mockProps} />);
    assertMenuClosed(wrapper);
  });

  it('should let the user open the menu by clicking on the control', () => {
    const wrapper = mount(<Dropdown {...mockProps} />);
    openMenu(wrapper);
    assertMenuOpen(wrapper, mockProps);
  });

  it('should render with strings as items', () => {
    const wrapper = mount(<Dropdown {...mockProps} items={['zar', 'doz']} />);
    openMenu(wrapper);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render custom item components', () => {
    const wrapper = mount(<Dropdown {...mockProps} />);
    wrapper.setProps({
      itemToElement: (item) => <div className="mock-item">{item.label}</div>,
    });
    openMenu(wrapper);
    expect(wrapper).toMatchSnapshot();
  });

  describe('title', () => {
    let wrapper;
    let renderedLabel;

    beforeEach(() => {
      wrapper = mount(<Dropdown titleText="Email Input" {...mockProps} />);
      renderedLabel = wrapper.find('label[className="bx--label"]');
    });

    it('renders a title', () => {
      expect(renderedLabel.length).toBe(1);
    });

    it('has the expected classes', () => {
      expect(renderedLabel.hasClass(`${prefix}--label`)).toEqual(true);
    });

    it('should set title as expected', () => {
      expect(renderedLabel.text()).toEqual('Email Input');
    });
  });

  describe('helper', () => {
    it('renders a helper', () => {
      const wrapper = mount(
        <Dropdown helperText="Email Input" {...mockProps} />
      );
      const renderedHelper = wrapper.find(`.${prefix}--form__helper-text`);
      expect(renderedHelper.length).toEqual(1);
    });

    it('renders children as expected', () => {
      const wrapper = mount(
        <Dropdown
          helperText={
            <span>
              This helper text has <a href="/">a link</a>.
            </span>
          }
          {...mockProps}
        />
      );
      const renderedHelper = wrapper.find(`.${prefix}--form__helper-text`);
      expect(renderedHelper.props().children).toEqual(
        <span>
          This helper text has <a href="/">a link</a>.
        </span>
      );
    });

    it('should set helper text as expected', () => {
      const wrapper = mount(<Dropdown {...mockProps} />);
      wrapper.setProps({ helperText: 'Helper text' });
      const renderedHelper = wrapper.find(`.${prefix}--form__helper-text`);
      expect(renderedHelper.text()).toEqual('Helper text');
    });
  });

  it('should specify light version as expected', () => {
    const wrapper = mount(<Dropdown {...mockProps} />);
    expect(wrapper.props().light).toEqual(false);
    wrapper.setProps({ light: true });
    expect(wrapper.props().light).toEqual(true);
  });

  it('should let the user select an option by clicking on the option node', () => {
    const wrapper = mount(<Dropdown {...mockProps} />);
    openMenu(wrapper);
    findMenuItemNode(wrapper, 0).simulate('click');
    expect(mockProps.onChange).toHaveBeenCalledTimes(1);
    expect(mockProps.onChange).toHaveBeenCalledWith({
      selectedItem: mockProps.items[0],
    });
    assertMenuClosed(wrapper);

    mockProps.onChange.mockClear();

    openMenu(wrapper);
    findMenuItemNode(wrapper, 1).simulate('click');
    expect(mockProps.onChange).toHaveBeenCalledTimes(1);
    expect(mockProps.onChange).toHaveBeenCalledWith({
      selectedItem: mockProps.items[1],
    });
  });

  describe('should display initially selected item found in `initialSelectedItem`', () => {
    it('using an object type for the `initialSelectedItem` prop', () => {
      const wrapper = mount(
        <Dropdown {...mockProps} initialSelectedItem={mockProps.items[0]} />
      );

      expect(wrapper.find(`span.${prefix}--list-box__label`).text()).toEqual(
        mockProps.items[0].label
      );
    });

    it('using a string type for the `initialSelectedItem` prop', () => {
      // Replace the 'items' property in mockProps with a list of strings
      mockProps = {
        ...mockProps,
        items: ['1', '2', '3'],
      };

      const wrapper = mount(
        <Dropdown {...mockProps} initialSelectedItem={mockProps.items[1]} />
      );

      expect(wrapper.find(`span.${prefix}--list-box__label`).text()).toEqual(
        mockProps.items[1]
      );
    });
  });

  describe('Component API', () => {
    afterEach(cleanup);

    it('should accept a `ref` for the underlying button element', () => {
      const ref = React.createRef();
      render(<Dropdown {...mockProps} ref={ref} />);
      expect(ref.current.getAttribute('aria-haspopup')).toBe('listbox');
    });
  });
});

describe('DropdownSkeleton', () => {
  describe('Renders as expected', () => {
    const wrapper = shallow(<DropdownSkeleton size="sm" />);

    it('Has the expected classes', () => {
      expect(wrapper.hasClass(`${prefix}--skeleton`)).toEqual(true);
      expect(wrapper.hasClass(`${prefix}--dropdown-v2`)).toEqual(true);
      expect(wrapper.hasClass(`${prefix}--list-box--sm`)).toEqual(true);
    });
  });
});
