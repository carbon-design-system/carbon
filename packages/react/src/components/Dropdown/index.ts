/**
 * Copyright IBM Corp. 2016, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Dropdown, { OnChangeData } from './Dropdown';

export type { OnChangeData };
export { Dropdown };
export { default as DropdownSkeleton } from './Dropdown.Skeleton';

export default Dropdown;
