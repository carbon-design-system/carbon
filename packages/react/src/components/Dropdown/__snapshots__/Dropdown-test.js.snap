// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render 1`] = `
<Dropdown
  direction="bottom"
  disabled={false}
  helperText=""
  id="test-dropdown"
  itemToElement={null}
  itemToString={[Function]}
  items={
    Array [
      Object {
        "id": "id-0",
        "label": "Item 0",
        "value": 0,
      },
      Object {
        "id": "id-1",
        "label": "Item 1",
        "value": 1,
      },
      Object {
        "id": "id-2",
        "label": "Item 2",
        "value": 2,
      },
      Object {
        "id": "id-3",
        "label": "Item 3",
        "value": 3,
      },
      Object {
        "id": "id-4",
        "label": "Item 4",
        "value": 4,
      },
    ]
  }
  label="input"
  light={false}
  onChange={[MockFunction]}
  placeholder="Filter..."
  titleText=""
  type="default"
>
  <div
    className="bx--dropdown__wrapper bx--list-box__wrapper"
    placeholder="Filter..."
  >
    <ListBox
      className="bx--dropdown"
      disabled={false}
      id="test-dropdown"
      isOpen={false}
      light={false}
      type="default"
    >
      <div
        className="bx--dropdown bx--list-box"
        id="test-dropdown"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <button
          aria-disabled={false}
          aria-expanded={false}
          aria-haspopup="listbox"
          aria-labelledby="downshift-0-label downshift-0-toggle-button"
          className="bx--list-box__field"
          disabled={false}
          id="downshift-0-toggle-button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span
            className="bx--list-box__label"
          >
            input
          </span>
          <ListBoxMenuIcon
            isOpen={false}
            translateWithId={[Function]}
          >
            <div
              className="bx--list-box__menu-icon"
            >
              <ForwardRef(ChevronDown16)
                aria-label="Open menu"
                name="chevron--down"
              >
                <Icon
                  aria-label="Open menu"
                  fill="currentColor"
                  height={16}
                  name="chevron--down"
                  preserveAspectRatio="xMidYMid meet"
                  viewBox="0 0 16 16"
                  width={16}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <svg
                    aria-label="Open menu"
                    fill="currentColor"
                    focusable="false"
                    height={16}
                    name="chevron--down"
                    preserveAspectRatio="xMidYMid meet"
                    role="img"
                    viewBox="0 0 16 16"
                    width={16}
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                    />
                    <title>
                      Open menu
                    </title>
                  </svg>
                </Icon>
              </ForwardRef(ChevronDown16)>
            </div>
          </ListBoxMenuIcon>
        </button>
        <ListBoxMenu
          aria-labelledby="downshift-0-label"
          id="downshift-0-menu"
          onBlur={[Function]}
          onKeyDown={[Function]}
          onMouseLeave={[Function]}
          role="listbox"
          tabIndex={-1}
        >
          <div
            aria-labelledby="downshift-0-label"
            className="bx--list-box__menu"
            id="downshift-0-menu"
            onBlur={[Function]}
            onKeyDown={[Function]}
            onMouseLeave={[Function]}
            role="listbox"
            tabIndex={-1}
          />
        </ListBoxMenu>
      </div>
    </ListBox>
  </div>
</Dropdown>
`;

exports[`Dropdown should render custom item components 1`] = `
<Dropdown
  direction="bottom"
  disabled={false}
  helperText=""
  id="test-dropdown"
  itemToElement={[Function]}
  itemToString={[Function]}
  items={
    Array [
      Object {
        "id": "id-0",
        "label": "Item 0",
        "value": 0,
      },
      Object {
        "id": "id-1",
        "label": "Item 1",
        "value": 1,
      },
      Object {
        "id": "id-2",
        "label": "Item 2",
        "value": 2,
      },
      Object {
        "id": "id-3",
        "label": "Item 3",
        "value": 3,
      },
      Object {
        "id": "id-4",
        "label": "Item 4",
        "value": 4,
      },
    ]
  }
  label="input"
  light={false}
  onChange={[MockFunction]}
  placeholder="Filter..."
  titleText=""
  type="default"
>
  <div
    className="bx--dropdown__wrapper bx--list-box__wrapper"
    placeholder="Filter..."
  >
    <ListBox
      className="bx--dropdown bx--dropdown--open"
      disabled={false}
      id="test-dropdown"
      isOpen={true}
      light={false}
      type="default"
    >
      <div
        className="bx--dropdown bx--dropdown--open bx--list-box bx--list-box--expanded"
        id="test-dropdown"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <button
          aria-disabled={false}
          aria-expanded={true}
          aria-haspopup="listbox"
          aria-labelledby="downshift-6-label downshift-6-toggle-button"
          className="bx--list-box__field"
          disabled={false}
          id="downshift-6-toggle-button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span
            className="bx--list-box__label"
          >
            input
          </span>
          <ListBoxMenuIcon
            isOpen={true}
            translateWithId={[Function]}
          >
            <div
              className="bx--list-box__menu-icon bx--list-box__menu-icon--open"
            >
              <ForwardRef(ChevronDown16)
                aria-label="Close menu"
                name="chevron--down"
              >
                <Icon
                  aria-label="Close menu"
                  fill="currentColor"
                  height={16}
                  name="chevron--down"
                  preserveAspectRatio="xMidYMid meet"
                  viewBox="0 0 16 16"
                  width={16}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <svg
                    aria-label="Close menu"
                    fill="currentColor"
                    focusable="false"
                    height={16}
                    name="chevron--down"
                    preserveAspectRatio="xMidYMid meet"
                    role="img"
                    viewBox="0 0 16 16"
                    width={16}
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                    />
                    <title>
                      Close menu
                    </title>
                  </svg>
                </Icon>
              </ForwardRef(ChevronDown16)>
            </div>
          </ListBoxMenuIcon>
        </button>
        <ListBoxMenu
          aria-labelledby="downshift-6-label"
          id="downshift-6-menu"
          onBlur={[Function]}
          onKeyDown={[Function]}
          onMouseLeave={[Function]}
          role="listbox"
          tabIndex={-1}
        >
          <div
            aria-labelledby="downshift-6-label"
            className="bx--list-box__menu"
            id="downshift-6-menu"
            onBlur={[Function]}
            onKeyDown={[Function]}
            onMouseLeave={[Function]}
            role="listbox"
            tabIndex={-1}
          >
            <ForwardRef(ListBoxMenuItem)
              aria-selected="false"
              id="downshift-6-item-0"
              isActive={false}
              isHighlighted={false}
              key="downshift-6-item-0"
              onClick={[Function]}
              onMouseMove={[Function]}
              role="option"
            >
              <div
                aria-selected="false"
                className="bx--list-box__menu-item"
                id="downshift-6-item-0"
                onClick={[Function]}
                onMouseMove={[Function]}
                role="option"
              >
                <div
                  className="bx--list-box__menu-item__option"
                >
                  <itemToElement
                    id="id-0"
                    key="downshift-6-item-0"
                    label="Item 0"
                    value={0}
                  >
                    <div
                      className="mock-item"
                    >
                      Item 0
                    </div>
                  </itemToElement>
                </div>
              </div>
            </ForwardRef(ListBoxMenuItem)>
            <ForwardRef(ListBoxMenuItem)
              aria-selected="false"
              id="downshift-6-item-1"
              isActive={false}
              isHighlighted={false}
              key="downshift-6-item-1"
              onClick={[Function]}
              onMouseMove={[Function]}
              role="option"
            >
              <div
                aria-selected="false"
                className="bx--list-box__menu-item"
                id="downshift-6-item-1"
                onClick={[Function]}
                onMouseMove={[Function]}
                role="option"
              >
                <div
                  className="bx--list-box__menu-item__option"
                >
                  <itemToElement
                    id="id-1"
                    key="downshift-6-item-1"
                    label="Item 1"
                    value={1}
                  >
                    <div
                      className="mock-item"
                    >
                      Item 1
                    </div>
                  </itemToElement>
                </div>
              </div>
            </ForwardRef(ListBoxMenuItem)>
            <ForwardRef(ListBoxMenuItem)
              aria-selected="false"
              id="downshift-6-item-2"
              isActive={false}
              isHighlighted={false}
              key="downshift-6-item-2"
              onClick={[Function]}
              onMouseMove={[Function]}
              role="option"
            >
              <div
                aria-selected="false"
                className="bx--list-box__menu-item"
                id="downshift-6-item-2"
                onClick={[Function]}
                onMouseMove={[Function]}
                role="option"
              >
                <div
                  className="bx--list-box__menu-item__option"
                >
                  <itemToElement
                    id="id-2"
                    key="downshift-6-item-2"
                    label="Item 2"
                    value={2}
                  >
                    <div
                      className="mock-item"
                    >
                      Item 2
                    </div>
                  </itemToElement>
                </div>
              </div>
            </ForwardRef(ListBoxMenuItem)>
            <ForwardRef(ListBoxMenuItem)
              aria-selected="false"
              id="downshift-6-item-3"
              isActive={false}
              isHighlighted={false}
              key="downshift-6-item-3"
              onClick={[Function]}
              onMouseMove={[Function]}
              role="option"
            >
              <div
                aria-selected="false"
                className="bx--list-box__menu-item"
                id="downshift-6-item-3"
                onClick={[Function]}
                onMouseMove={[Function]}
                role="option"
              >
                <div
                  className="bx--list-box__menu-item__option"
                >
                  <itemToElement
                    id="id-3"
                    key="downshift-6-item-3"
                    label="Item 3"
                    value={3}
                  >
                    <div
                      className="mock-item"
                    >
                      Item 3
                    </div>
                  </itemToElement>
                </div>
              </div>
            </ForwardRef(ListBoxMenuItem)>
            <ForwardRef(ListBoxMenuItem)
              aria-selected="false"
              id="downshift-6-item-4"
              isActive={false}
              isHighlighted={false}
              key="downshift-6-item-4"
              onClick={[Function]}
              onMouseMove={[Function]}
              role="option"
            >
              <div
                aria-selected="false"
                className="bx--list-box__menu-item"
                id="downshift-6-item-4"
                onClick={[Function]}
                onMouseMove={[Function]}
                role="option"
              >
                <div
                  className="bx--list-box__menu-item__option"
                >
                  <itemToElement
                    id="id-4"
                    key="downshift-6-item-4"
                    label="Item 4"
                    value={4}
                  >
                    <div
                      className="mock-item"
                    >
                      Item 4
                    </div>
                  </itemToElement>
                </div>
              </div>
            </ForwardRef(ListBoxMenuItem)>
          </div>
        </ListBoxMenu>
      </div>
    </ListBox>
  </div>
</Dropdown>
`;

exports[`Dropdown should render with strings as items 1`] = `
<Dropdown
  direction="bottom"
  disabled={false}
  helperText=""
  id="test-dropdown"
  itemToElement={null}
  itemToString={[Function]}
  items={
    Array [
      "zar",
      "doz",
    ]
  }
  label="input"
  light={false}
  onChange={[MockFunction]}
  placeholder="Filter..."
  titleText=""
  type="default"
>
  <div
    className="bx--dropdown__wrapper bx--list-box__wrapper"
    placeholder="Filter..."
  >
    <ListBox
      className="bx--dropdown bx--dropdown--open"
      disabled={false}
      id="test-dropdown"
      isOpen={true}
      light={false}
      type="default"
    >
      <div
        className="bx--dropdown bx--dropdown--open bx--list-box bx--list-box--expanded"
        id="test-dropdown"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <button
          aria-disabled={false}
          aria-expanded={true}
          aria-haspopup="listbox"
          aria-labelledby="downshift-4-label downshift-4-toggle-button"
          className="bx--list-box__field"
          disabled={false}
          id="downshift-4-toggle-button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span
            className="bx--list-box__label"
          >
            input
          </span>
          <ListBoxMenuIcon
            isOpen={true}
            translateWithId={[Function]}
          >
            <div
              className="bx--list-box__menu-icon bx--list-box__menu-icon--open"
            >
              <ForwardRef(ChevronDown16)
                aria-label="Close menu"
                name="chevron--down"
              >
                <Icon
                  aria-label="Close menu"
                  fill="currentColor"
                  height={16}
                  name="chevron--down"
                  preserveAspectRatio="xMidYMid meet"
                  viewBox="0 0 16 16"
                  width={16}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <svg
                    aria-label="Close menu"
                    fill="currentColor"
                    focusable="false"
                    height={16}
                    name="chevron--down"
                    preserveAspectRatio="xMidYMid meet"
                    role="img"
                    viewBox="0 0 16 16"
                    width={16}
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                    />
                    <title>
                      Close menu
                    </title>
                  </svg>
                </Icon>
              </ForwardRef(ChevronDown16)>
            </div>
          </ListBoxMenuIcon>
        </button>
        <ListBoxMenu
          aria-labelledby="downshift-4-label"
          id="downshift-4-menu"
          onBlur={[Function]}
          onKeyDown={[Function]}
          onMouseLeave={[Function]}
          role="listbox"
          tabIndex={-1}
        >
          <div
            aria-labelledby="downshift-4-label"
            className="bx--list-box__menu"
            id="downshift-4-menu"
            onBlur={[Function]}
            onKeyDown={[Function]}
            onMouseLeave={[Function]}
            role="listbox"
            tabIndex={-1}
          >
            <ForwardRef(ListBoxMenuItem)
              aria-selected="false"
              id="downshift-4-item-0"
              isActive={false}
              isHighlighted={false}
              key="downshift-4-item-0"
              onClick={[Function]}
              onMouseMove={[Function]}
              role="option"
            >
              <div
                aria-selected="false"
                className="bx--list-box__menu-item"
                id="downshift-4-item-0"
                onClick={[Function]}
                onMouseMove={[Function]}
                role="option"
              >
                <div
                  className="bx--list-box__menu-item__option"
                >
                  zar
                </div>
              </div>
            </ForwardRef(ListBoxMenuItem)>
            <ForwardRef(ListBoxMenuItem)
              aria-selected="false"
              id="downshift-4-item-1"
              isActive={false}
              isHighlighted={false}
              key="downshift-4-item-1"
              onClick={[Function]}
              onMouseMove={[Function]}
              role="option"
            >
              <div
                aria-selected="false"
                className="bx--list-box__menu-item"
                id="downshift-4-item-1"
                onClick={[Function]}
                onMouseMove={[Function]}
                role="option"
              >
                <div
                  className="bx--list-box__menu-item__option"
                >
                  doz
                </div>
              </div>
            </ForwardRef(ListBoxMenuItem)>
          </div>
        </ListBoxMenu>
      </div>
    </ListBox>
  </div>
</Dropdown>
`;
